text/0000-generic-dependent-consts.md

```diff
@@ -0,0 +1,291 @@
+- Feature Name: generic_dependent_consts
+- Start Date: 2015-03-28
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Allow consts declared in functions to depend on function type parameters. Since
+this raises some issues in match and type checks, use an conservative approach
+that allows only minimal use of such consts when a constant expression is
+required.
+
+# Motivation
+
+Consider the following line of code in a non-generic context:
+
+```rust
+let a: [u8; T::N + T::N] = [0u8; 2*T::N];
+```
+
+In the current associated consts implementation (modulo some temporary wrinkles
+involving constants associated to traits), this will readily type-check. Since
+the type `T` must be known in non-generic code, `T::N` will evaluate to some
+value, say `16`, in which case it is clear that both sides of the assignment
+involve arrays of size `32`.
+
+However, in generic code, if `T` is a type parameter, we are asking the type
```

Comment:
> I guess I don't really understand what the difference is between your case 2 [...] and having a C++ style static_assert inside the function body

Three differences I can think of:

 1. Constraints in where clauses interact across function boundaries. If I had some function `funky_fn<I, J>` that called some other function (that called another function...) that ultimately called your `bar` as `bar::<I, J>()`, then I might need to also specify `I::VALUE == J::VALUE` on `funky_fn` (and maintain this condition on all functions in between). As I mentioned, the downside is that this is more verbose. The upside is that this forces the restriction on `funky_fn` arguments to be specified up front.

    With static asserts, I might not know that there was a restriction on `funky_fn` until I tried calling it the wrong way, and then the error message would come from all the way down in `bar`, which has completely different names for its type parameters, and which might be an undocumented implementation detail that I'm not even supposed to have to deal with. This is one contributing factor to puzzling template error messages you sometimes get from C++ compilers.

 2. When the constraint is specified as part of the function signature, the compiler can tell that `A::VALUE == 2` and `A::VALUE == 3` are non-overlapping during coherence checking, and thus allow two implementations that differ only with respect to these constraints. This touches on function specialization, but I just want to clarify that this is something you get even if constants are treated just like types are now, without any of the things proposed for future improvements in specialization.

 3. If you put a constraint on a method's signature in a trait, it can be taken into account by users who have no idea which implementations of that method they could get in a piece of generic code. Depending on how we want to approach object safety, we could also have trait objects specified in types like `Box<Trait<CONST=3>>`. Sometimes you want to talk about constraints on a trait member even when you don't know what the actual implementation of the trait is at compile time.